### PR TITLE
feat(cfw): add resource CFW capture

### PR DESCRIPTION
**Pull request by profoundwu**
**What this PR does / why we need it**:
add resource CFW capture
**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
1. add resource CFW capture
```

## PR Checklist

* [x] Tests added/passed.

```
make testacc TEST="./huaweicloud/services/acceptance/cfw" TESTARGS="-run TestAccCapture_basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cfw -v -run TestAccCapture_basic -timeout 360m -parallel 1
=== RUN   TestAccCapture_basic
=== PAUSE TestAccCapture_basic
=== CONT  TestAccCapture_basic
--- PASS: TestAccCapture_basic (45.24s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cfw       45.299s

```

* [x] Documentation updated.
* [x] Schema updated.
* [x] CheckDeleted.

  - **a. During query operation (Read Context)**
    aa. Resource not found
   image

    
  - **b. During delete/disassociate/unbind operation (Delete Context)**
    ba. Resource not found
   image

### CODE REVIEW

File: docs/resources/cfw_capture.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
+---
+subcategory: "Cloud Firewall (CFW)"
+layout: "huaweicloud"
+page_title: "HuaweiCloud: huaweicloud_cfw_capture"
+description: ""
+---
```
Comment (ruwenqiang123) [inline docs/resources/cfw_capture.md]:
```suggestion
description: -|
  Manages a CFW capture resource within HuaweiCloud.
---
```

File: docs/resources/cfw_capture.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  
+  destination {
+    address = "1.1.1.1"
+    address_type = 0
+  }
+
+  source {
+    address = "2.2.2.2"
+    address_type = 0
+  }
```
Comment (ruwenqiang123) [inline docs/resources/cfw_capture.md]:
```suggestion
  source {
    address      = "2.2.2.2"
    address_type = 0
  }
```

File: docs/resources/cfw_capture.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+resource huaweicloud_cfw_capture "test" {
+  fw_instance_id = var.fw_instance_id
+  name           = var.name
+  duration       = var.duration
+  max_packets    = var.max_packets
+  
+  destination {
+    address = "1.1.1.1"
+    address_type = 0
+  }
```
Comment (ruwenqiang123) [inline docs/resources/cfw_capture.md]:
```suggestion
  destination {
    address      = "1.1.1.1"
    address_type = 0
  }
```

File: docs/resources/cfw_capture.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+## Timeouts
+
+This resource provides the following timeouts configuration options:
+
+* `update` - Default is 5 minutes.
+* `delete` - Default is 5 minutes.
+
+## Import
+
+The capture can be imported using `fw_instance_id`, `name`, separated by slashes, e.g.
```
Comment (ruwenqiang123) [inline docs/resources/cfw_capture.md]:
```suggestion
The capture can be imported using `fw_instance_id`, `name`, separated by  a slashe (/), e.g.
```

File: docs/resources/cfw_capture.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+The capture can be imported using `fw_instance_id`, `name`, separated by a slash, e.g.
+
+```bash
+$ terraform import huaweicloud_cfw_capture.test <fw_instance_id>/<name>
+```
+
+Note that the imported state may not be identical to your resource definition, due to some attributes missing from the
+API response, security or some other reason.
+The missing attributes is `stop_capture`. It is generally recommended running `terraform plan` after importing the resource.
+You can then decide if changes should be applied to the instance, or the resource definition should be updated to
```
Comment (houpeng80) [inline docs/resources/cfw_capture.md]:
```suggestion
You can then decide if changes should be applied to the  capture, or the resource definition should be updated to
```

File: docs/resources/cfw_capture.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+The following arguments are supported:
+
+* `region` - (Optional, String, ForceNew) Specifies the region in which to create the resource.
+  If omitted, the provider-level region will be used.
+  Changing this creates a new resource.
+
+* `fw_instance_id` - (Required, String, NonUpdatable) Specifies the ID of the firewall instance.
+
+* `name` - (Required, String, NonUpdatable) Specifies the capture name.
```
Comment (houpeng80) [inline docs/resources/cfw_capture.md]:
```suggestion
* `name` - (Required, String, NonUpdatable) Specifies the capture task name.
```
and fix other places

File: docs/resources/cfw_capture.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+<a name="Source"></a>
+The `source` block supports:
+
+* `address` - (Required, String, NonUpdatable) Specifies the address.
+
+* `address_type` - (Required, Int, NonUpdatable) Specifies the address type.
+  The valid values are:
+  + **0**: indicates IPv4;
+  + **1**: indicates IPv6.
```
Comment (saf3dfsa) [inline docs/resources/cfw_capture.md]: Destination and Source blocks are same, maybe you can name it as `addresses` block.

File: huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_capture_test.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/cfw"
+)
+
+func getCaptureResourceFunc(cfg *config.Config, state *terraform.ResourceState) (interface{}, error) {
+	region := acceptance.HW_REGION_NAME
+	product := "cfw"
+
+	client, err := cfg.NewServiceClient(product, region)
+	if err != nil {
+		return nil, fmt.Errorf("error creating CFW Client: %s", err)
```
Comment (saf3dfsa) [inline huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_capture_test.go]:
```suggestion
		return nil, fmt.Errorf("error creating CFW client: %s", err)
```

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_capture.go [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+	opt.JSONBody = utils.RemoveNil(buildCreateCaptureBodyParams(d))
+	resp, err := client.Request("POST", path, &opt)
+	if err != nil {
+		return diag.Errorf("error creating capture: %s", err)
+	}
+
+	_, err = utils.FlattenResponse(resp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
```
Comment (houpeng80) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_capture.go:206]: it is useless

Comment (saf3dfsa) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_capture.go]:
```suggestion
	_, err := client.Request("POST", path, &opt)
	if err != nil {
		return diag.Errorf("error creating capture: %s", err)
	}
```

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_capture.go [hunk 10]
```diff
@@ -0,0 +1,498 @@
...
+
+		if len(curArray) == 0 {
+			return nil, golangsdk.ErrDefault404{}
+		}
+
+		findCaptureStr := fmt.Sprintf("[?name=='%s']|[0]", name)
+		capture := utils.PathSearch(findCaptureStr, curJson, nil)
+		if capture != nil {
+			return capture, nil
+		}
```
Comment (saf3dfsa) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_capture.go]: you can merge it into one pathSearch

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_capture.go [hunk 11]
```diff
@@ -0,0 +1,498 @@
...
+
+		offset += 100
```
Comment (saf3dfsa) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_capture.go]: the response has `total`, it's no need to search the next page if the datas are less than 100.

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_capture.go [hunk 12]
```diff
@@ -0,0 +1,498 @@
...
+		if err != nil {
+			log.Printf("[DEBUG] failed to stop capture: %#v", err)
+		}
+	}
+
+	name := d.Get("name").(string)
+	instanceID := d.Get("fw_instance_id").(string)
+	err = waitCaptureTaskStopped(ctx, client, name, instanceID, d.Timeout(schema.TimeoutDelete))
+	if err != nil {
+		return common.CheckDeletedDiag(d, parseError(err), "error deleting capture")
```
Comment (saf3dfsa) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_capture.go]:
```suggestion
		return common.CheckDeletedDiag(d, parseError(err), ""error waiting for CFW capture task stop")
```

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_capture.go [hunk 13]
```diff
@@ -0,0 +1,498 @@
...
+	}
```
Comment (houpeng80) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_capture.go]: move it to line 455? if the task is no need to stop, then no need to wait

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_capture.go [hunk 14]
```diff
@@ -0,0 +1,498 @@
...
+	// Call the detail API to check if the resource is actually deleted.
+	_, err = GetCapture(client, name, instanceID)
+
+	return common.CheckDeletedDiag(d, parseError(err), "error deleting capture")
+}
+
+func resourceCaptureImportState(_ context.Context, d *schema.ResourceData, _ interface{}) ([]*schema.ResourceData, error) {
+	parts := strings.SplitN(d.Id(), "/", 2)
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("invalid format specified for import id, must be <fw_instance_id>/<name>")
```
Comment (saf3dfsa) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_capture.go]:
```suggestion
		return nil, fmt.Errorf("invalid format specified for import ID, must be <fw_instance_id>/<name>")
```

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_capture_task.go [hunk 15]
```diff
@@ -0,0 +1,494 @@
...
+
+	name := d.Get("name").(string)
+	instanceID := d.Get("fw_instance_id").(string)
+	capture, err := GetCaptureTask(client, name, instanceID)
+	if err != nil {
+		return common.CheckDeletedDiag(d, parseError(err), "error retrieving capture task")
+	}
+
+	id := utils.PathSearch("task_id", capture, "")
+	d.SetId(id.(string))
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_capture_task.go]: why not use name as the id, and save task_id as an attribute

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_capture_task.go [hunk 16]
```diff
@@ -0,0 +1,494 @@
...
+
+	// Deletion API call does not guranatee the deletion of the resource.
+	// Call the detail API to check if the resource is actually deleted.
+	_, err = GetCaptureTask(client, name, instanceID)
+
+	return common.CheckDeletedDiag(d, parseError(err), "error deleting capture task")
+}
+
+func resourceCaptureTaskImportState(_ context.Context, d *schema.ResourceData, _ interface{}) ([]*schema.ResourceData, error) {
+	parts := strings.SplitN(d.Id(), "/", 2)
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_capture_task.go]:
```suggestion
	parts := strings.Split(d.Id(), "/")
```

**Pull request merged**